File: 105_construct_binary_tree_from_preorder_and_inorder_traversal.rb
=begin
Given preorder and inorder traversal of a tree, construct the binary tree.

Note:
You may assume that duplicates do not exist in the tree.

For example, given

preorder = [3,9,20,15,7]
inorder = [9,3,15,20,7]
Return the following binary tree:

    3
   / \
  9  20
    /  \
   15   7
=end

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {Integer[]} preorder
# @param {Integer[]} inorder
# @return {TreeNode}
def build_tree(preorder, inorder)
    @idx_map = {}
    @pre_idx = 0
    @preorder = preorder
    @inorder = inorder
    inorder.each_with_index do |val, i|
        @idx_map[val] = i
    end
    
    return helper(0, inorder.length)
end

def helper(in_left, in_right)
    if in_left == in_right
        return nil
    end
    
    root_val = @preorder[@pre_idx]
    root = TreeNode.new(root_val)
    
    index = @idx_map[root_val]
    @pre_idx += 1
    root.left = helper(in_left, index)
    root.right = helper(index + 1, in_right)
    
    return root
end

File: 105_construct_binary_tree_from_preorder_and_inorder_traversal.go
/*
Given two integer arrays preorder and inorder where preorder is the preorder traversal of a binary tree and inorder is the inorder traversal of the same tree, construct and return the binary tree.

 

Example 1:


Input: preorder = [3,9,20,15,7], inorder = [9,3,15,20,7]
Output: [3,9,20,null,null,15,7]
Example 2:

Input: preorder = [-1], inorder = [-1]
Output: [-1]
 

Constraints:

1 <= preorder.length <= 3000
inorder.length == preorder.length
-3000 <= preorder[i], inorder[i] <= 3000
preorder and inorder consist of unique values.
Each value of inorder also appears in preorder.
preorder is guaranteed to be the preorder traversal of the tree.
inorder is guaranteed to be the inorder traversal of the tree.
*/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(preorder []int, inorder []int) *TreeNode {
    if len(preorder) == 0 || len(inorder) == 0 { return nil }
    return helper(&preorder, &inorder, 0, 0, len(preorder)-1)
}

func helper(preorder, inorder *[]int, preStart, inStart, inEnd int) *TreeNode {
    if preStart > len(*preorder) || inStart > inEnd {
        return nil
    }
    current := &TreeNode{Val: (*preorder)[preStart]}
    i := inStart
    for i <= inEnd {
        if (*inorder)[i] == (*preorder)[preStart] { break }
        i++
    }
    current.Left = helper(preorder, inorder, preStart+1, inStart, i-1)
    current.Right = helper(preorder, inorder, preStart+(i-inStart+1), i+1, inEnd)
    return current
}

